middleware/grpc_middleware_test.go

```diff
@@ -0,0 +1,310 @@
+					Authorizer:     rule.Handler{Handler: "allow"},
+					Mutators:       []rule.Handler{{Handler: "noop"}},
+				}},
+			},
+			assert: assertErrDenied,
+		},
+		{
+			name: "should succeed when correct bearer token is supplied",
+			rules: map[configuration.MatchingStrategy][]rule.Rule{
+				configuration.Regexp: {{
+					Match: &rule.MatchGRPC{
+						Authority:  "<([0-9a-zA-Z\\-]+)>.apis.ory.sh",
+						FullMethod: "grpc.testing.TestService/EmptyCall",
+					},
+					Authenticators: []rule.Handler{{Handler: "bearer_token"}},
+					Authorizer:     rule.Handler{Handler: "allow"},
+					Mutators:       []rule.Handler{{Handler: "noop"}},
+				}},
+				configuration.Glob: {{
+					Match: &rule.MatchGRPC{
+						Authority:  "<**>.apis.ory.sh",
+						FullMethod: "grpc.testing.TestService/EmptyCall",
+					},
+					Authenticators: []rule.Handler{{Handler: "bearer_token"}},
+					Authorizer:     rule.Handler{Handler: "allow"},
+					Mutators:       []rule.Handler{{Handler: "noop"}},
+				}},
+			},
+			dialOpts: []grpc.DialOption{grpc.WithPerRPCCredentials(
+				testToken("correct token"))},
+			assert: assert.NoError,
+		},
+		{
+			name: "should fail when incorrect bearer token is supplied",
+			rules: map[configuration.MatchingStrategy][]rule.Rule{
+				configuration.Regexp: {{
+					Match: &rule.MatchGRPC{
+						Authority:  "<([0-9a-zA-Z\\-]+)>.apis.ory.sh",
+						FullMethod: "grpc.testing.TestService/EmptyCall",
+					},
+					Authenticators: []rule.Handler{{Handler: "bearer_token"}},
+					Authorizer:     rule.Handler{Handler: "allow"},
+					Mutators:       []rule.Handler{{Handler: "noop"}},
+				}},
+				configuration.Glob: {{
+					Match: &rule.MatchGRPC{
+						Authority:  "<**>.apis.ory.sh",
+						FullMethod: "grpc.testing.TestService/EmptyCall",
+					},
+					Authenticators: []rule.Handler{{Handler: "bearer_token"}},
+					Authorizer:     rule.Handler{Handler: "allow"},
+					Mutators:       []rule.Handler{{Handler: "noop"}},
+				}},
+			},
+			dialOpts: []grpc.DialOption{grpc.WithPerRPCCredentials(
+				testToken("incorrect token"))},
+			assert: assertErrDenied,
+		},
+	}
+
+	strategies := []configuration.MatchingStrategy{configuration.Regexp, configuration.Glob}
+
+	for _, tc := range cases {
+		t.Run("case="+tc.name, func(t *testing.T) {
+			client := testClient(t, l, tc.dialOpts...)
+			for _, s := range strategies {
+				t.Run("strategy="+string(s), func(t *testing.T) {
+
+					require.NoError(t, reg.RuleRepository().
+						SetMatchingStrategy(ctx, s))
+					reg.RuleRepository().(*rule.RepositoryMemory).
+						WithRules(tc.rules[s])
+
+					_, err := client.EmptyCall(ctx, &grpcTesting.Empty{})
+					tc.assert(t, err)
+
+					_, err = client.UnaryCall(ctx, &grpcTesting.SimpleRequest{})
+					assertErrDenied(t, err)
+
+					stream, _ := client.StreamingOutputCall(ctx, &grpcTesting.StreamingOutputCallRequest{})
+					_, err = stream.Recv()
+					assertErrDenied(t, err)
```

Comment:
Right now this is out of scope of what the middleware can do (it will always deny streaming requests).

We don't need them right now so I didn't think too hard about whether the semantics translate nicely or if we need another field (now that we have the special `MatchGRPC` type).

WDYT? I would like to defer streaming to later when we understand the constrains better.

Reply:
OK makes sense, I just thought the streaming interceptor is already there and works? But it's fine by me to not have support for now.

Reply:
It's there, but are the semantics the same? For example, if streaming connections are open for a long time, do we want to re-validate the checks at some point?

F.e., we might want to close streaming connections if a user session is closed.

That's what I meant by semantics are still unclear for me.

Reply:
Yes, what you implemented right now does a check on every message submitted through the stream, which might get very slow (especially with `remote_json` or similar rules).